src/Microsoft.AspNet.Mvc.Razor/Buffer/RazorBufferTextWriter.cs

```diff
@@ -0,0 +1,166 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNet.Html.Abstractions;
+using Microsoft.AspNet.Mvc.Internal;
+
+namespace Microsoft.AspNet.Mvc.Razor.Buffer
+{
+    /// <summary>
+    /// A <see cref="HtmlTextWriter"/> that stores individual write operations as a sequence of
+    /// <see cref="string"/> and <see cref="IHtmlContent"/> instances.
+    /// </summary>
+    /// <remarks>
+    /// This is primarily designed to avoid creating large in-memory strings.
+    /// Refer to https://aspnetwebstack.codeplex.com/workitem/585 for more details.
+    /// </remarks>
+    public class RazorBufferTextWriter : HtmlTextWriter
+    {
+        private const int MaxCharToStringLength = 1024;
+
+        private readonly Encoding _encoding;
+        private readonly RazorBuffer _buffer;
+
+        public RazorBufferTextWriter(RazorBuffer buffer, Encoding encoding)
+        {
+            _buffer = buffer;
+            _encoding = encoding;
+        }
+
+        /// <inheritdoc />
+        public override Encoding Encoding
+        {
+            get { return _encoding; }
+        }
+
+        /// <summary>
+        /// Gets the content written to the writer as an <see cref="IHtmlContent"/>.
+        /// </summary>
+        public IHtmlContent Content => _buffer;
+
+        /// <inheritdoc />
+        public override void Write(char value)
+        {
+            Write(value.ToString());
+        }
+
+        /// <inheritdoc />
+        public override void Write(char[] buffer, int index, int count)
+        {
+            if (buffer == null)
+            {
+                throw new ArgumentNullException(nameof(buffer));
+            }
+
+            if (index < 0)
+            {
+                throw new ArgumentOutOfRangeException(nameof(index));
+            }
+            if (count < 0 || (buffer.Length - index < count))
+            {
+                throw new ArgumentOutOfRangeException(nameof(count));
+            }
+
+            while (count > 0)
+            {
+                // Split large char arrays into 1KB strings.
+                var currentCount = count;
+                if (MaxCharToStringLength < currentCount)
+                {
+                    currentCount = MaxCharToStringLength;
+                }
+
+                Write(new string(buffer, index, currentCount));
+                index += currentCount;
+                count -= currentCount;
+            }
+        }
+
+        /// <inheritdoc />
+        public override void Write(string value)
+        {
+            if (string.IsNullOrEmpty(value))
+            {
+                return;
+            }
+
+            _buffer.AppendHtml(value);
```

Comment:
That sounds like a big rathole. I'd suggest we wait for evidence that's a good idea.

Generally the long strings come from static razor content, so they aren't eligible for GC.